### DISCUSSION TITLE
Fire TV - start & exit of "mirroring mode"

### DISCUSSION THREAD

**Discussion by camelcamro** [PRProd/HA-Firemote#351] [Q&A] [answered]
### App Name

mirroring mode (fire tv 4k)

### Device Platform

Amazon Fire

### App Store Link and/or URL to .apk file if sideloaded

(system integrated)

### Application Details

normally, it is needed to go into settings menu, and then click on "duplicate screen" to activate

### Optional Comments

would be great, if that could be done with a single button and not doing it through complicate way through "settings menu" and so on
further more, I haven't found app: "Kodi" but that is another feature request, I guess

**Comment by PRProd** [comment, marked as answer]
@camelcamro ~

Wow!  This idea led me down quite a long road of issuing commands to see what happens... Here is what I found:

Unexpectedly, when my FireTV device is in Display Mirroring mode, it stops receiving ADB commands.  In fact, it seems that all ADB communication is severed while in that mode.  So far I've only tested this with my [Fire TV 4 Series (2021)](https://github.com/PRProd/HA-Firemote/wiki/Smart-TV---Fire-TV-4-Series-(2021)) so I'm hoping that other Fire TV devices won't have this same behavior.  The only way I was able to restore ADB/Firemote control after entering mirroring mode is to press the **back** button on my physical remote control.

With that surprising setback of not being able to issue an ADB command to interrogate my TV while it was in Display Mirroring mode, I had to do a lot of poking around, but eventually I found what I was looking for.  This command will start the Display Mirroring Mode
```
adb shell am start com.amazon.cast.sink/.DisplayMirroringSinkActivity
```

So... I can definitely add a function button that launches the display mirroring mode, but when you press it you might be stuck there without a "Firemote" way of getting back.

What are your thoughts about that?

**Comment by camelcamro** [comment]
wow ... you are amazing !!!
yes, that is working ...

and YES, there is no way bac via adb cmd, because you can't connect, as there is not any more "NETWORK" available.
because this cast method, is disabling network totally, and making a direct (WIFI) connection.
for more infos: https://en.wikipedia.org/wiki/Miracast

Technical details
Miracast is based on the peer-to-peer Wi-Fi Direct standard. It allows sending up to 1080p HD video (H.264 codec) and 5.1 surround sound (AAC and AC3 are optional codecs, mandated codec is linear pulse-code modulation – 16 bits 48 kHz 2 channels).[13] The connection is created via WPS and therefore is secured with WPA2. IPv4 is used on the Internet layer. On the transport layer, TCP or UDP are used. On the application layer, the stream is initiated and controlled via RTSP, RTP for the data transfer.


so, i tested your given adb cmd, and it was working to activate the mirroring mode ...

hmm, and now ? how to disable without a real (IR) firemote control ?
well, in my case, it is simple, as i can also record signal from the real physical fire-tv--remote and send IR/RF cmd through my "Broadlink Remote RM+" to DISABLE THE MIRRORING MODE ...

but without ? and only doing it via adb (if the network and adb server is disabled by MIRRORING MODE) - that is impossible as network/adb not available

so, the challenge is maybe, if the keys/events can be simulated ... not sure, at the moment ... need brain strorming ...

https://developer.amazon.com/docs/fire-tv/remote-input.html

on the other side - i can live with that dirty solution

**Comment by camelcamro** [comment]
in any case, i would not add these feature to a button, as it can also be done via a script and executing the adb comand.

adb shell am start com.amazon.cast.sink/.DisplayMirroringSinkActivity

not needed to add it with an extra button ...as long as it can not be disabled with another method ...
######
additionally, i hate on this miracast, that it is only a screen mirroring ... nothing else, and need to switch to that mode ...

i guess, there is a "AirReceiver" for fire tv stick... and on android and IOS supported by the "real google cast" and/or airplay by deault no apps needed.
and for windows, then there is the "AirSender" app ... thats it

**Comment by PRProd** [comment]
I'm going to convert this issue into a Q&A discussion just in case someone else has a similar request.  It's probably easier to find it that way.

**Comment by camelcamro** [comment]
I'M trying to keep it short - as it would be a long story ....
in my case - i only tested: Amazon Firestick 4K Max (2nd Gen 2023)

1.) internal built-in MIRRORING SYSTEM
 => only supports MIRACAST - and also block the whole internal network
 => MIRACAST is not supported by google anymore and removed from their own devices
 => Windows would support miracast out of the box, but is also very limted - based on miracast (see wiki miracast)

so, we would need an app on firestick, which supports much more: MIRACAST, CHROMECAST, DLNA, AIRPLAY, etc..
and should be not destroying the connection to the firestick (adb is needed to handle remote/HA-Firemote)

so, the solution i found for me is to replace the internal mirrowing system, which is a mess and totally limited - so: for me is/was:

=>
installing an APP on firestick:
=> AIRSENDER (very important see the developer in amazon stor, as there is another also existing)
https://www.amazon.de/AirScreen-AirPlay-Google-Cast-Miracast/dp/B07CPZ698R
it is:
.) easy to install via the normal AMAZON STORE
.) can run in background (starting as backfground service automatically was not possible for me til now, but possible to start by myself)
.) supports: AirPlay, Cast, Chromecast, Miracast, Cast-enabled, DLNA
.) free of charge on basic function - wich is sufficient
.) pro version would have a lot of more options if needed
.) apple - IOS and macOS (apple) - client devices supported (OUT IOF THE BOX)
.) android OS - (most) client devices supported (OUT IOF THE BOX)
.) windows OS - supported by 3rd party app: 
    "airparrot3" - is not freeware (~18 USD), maybe there is another freeware, not sure. for me, it is good as i have a lot of options to say what i want to mirror ..
    "AirMyPC" - is also not free and also around 18 USD (not tested)

and now, i only need to find out how i can start this app on firestick also through adb

for me, is that a much more better solution without any limitation.
